docs/proposals/parquet-storage.md

```diff
@@ -0,0 +1,128 @@
+- Rich file metadata and index, no local state like index header required to query the data, making it stateless
+- Advanced compression techniques reduce storage costs and improve query performance
+- Parallel processing friendly using Parquet Row Group
+
+There are other benefits of Parquet formats, but they are not directly related to the proposal:
+
+- Wide ecosystem and tooling support
+- Column pruning opportunity using projection pushdown
+
+## Out of Scope
+
+- Allow Ingester and Compactor to create Parquet files instead of TSDB blocks directly. This could be in the future roadmap but this proposal only focuses on converting and querying Parquet files.
+
+## Proposed Design
+
+### Components
+
+There are 2 new Cortex components/modules introduced in this design.
+
+#### 1. Parquet Converter
+
+Parquet Converter is a new component that converts TSDB blocks on object store to Parquet file format.
+
+It is similar to compactor, however, it only converts single block. The converted Parquet files will be stored in the same TSDB block folder so that the lifecycle of Parquet file will be managed together with the block.
+
+Only certain blocks can be configured to convert to Parquet file and it will be block duration based, for example we only convert if block duration is >= 12h.
+
+#### 2. Parquet Queryable
+
+Similar to the existing distributorQueryable and blockStorageQueryable, Parquet queryable is a queryable implementation which allows Cortex to query parquet files and can be used in both Cortex Querier and Ruler.
+
+If Parquet queryable is enabled, block storage queryable will be disabled and Cortex querier will not query Store Gateway anymore. But it still queries Ingesters.
+
+Cortex querier remains a stateless component when Parquet queryable is enabled.
+
+### Architecture
+
+```
+┌──────────┐    ┌─────────────┐    ┌──────────────┐
+│ Ingester │───>│   TSDB      │───>│   Parquet    │
+└──────────┘    │   Blocks    │    │  Converter   │
+                └─────────────┘    └──────────────┘
+                                          │
+                                          v
+┌──────────┐    ┌─────────────┐    ┌──────────────┐
+│  Query   │───>│   Parquet   │───>│    Parquet   │
+│ Frontend │    │   Querier   │    │    Files     │
+└──────────┘    └─────────────┘    └──────────────┘
+```
+
+### Data Format
+
+Following the current desgin of Cortex, each Parquet file contains at most 1 day of data.
+
+#### Schema Overview
+
+The Parquet format consists of two types of files:
+
+1. **Labels Parquet File**
+   - Each row represents a unique time series
+   - Each column corresponds to a label name (e.g., `__name__`, `label1`, ..., `labelN`)
+   - Row groups are sorted by `__name__` alphabetically in ascending order
+
+2. **Chunks Parquet File**
+   - Maintains row and row group order matching the Labels file
+   - Contains multiple chunk columns for time-series data. Each column covering a time range of chunks: 0-8h, 8h-16h, 16-24h.
```

Comment:
Could you clarify the benefit of chunks parquet file over the existing chunks file? The labels parquet file clearly has advantage of fetching all series for label matchers, but once we have the final list of series to fetch, I'm not sure how having chunks in parquet file will help with performance or memory utilization.

Reply:
It enables fetching chunks for 8h ( by default, configurable ) in groups. This reduces requests to object storage if you dont need to fetch chunks for all 24h of the block.

Reply:
I think the main benefit here is overfetch. We have seen quite bad overfetch on Store Gateway for chunks where fetched chunk bytes vs touched chunk bytes is 20:1.

There could be overfetch in Parquet as well but the smallest read unit in Parquet is page and the size is configurable. From our initial test the overfetch seems much better.

Parquet's compression and encoding is another nice addition as we able to see 30% of chunk size reduction.